include/alpaka/dev/DevGenericSycl.hpp

```diff
@@ -0,0 +1,165 @@
+{
+    namespace experimental
+    {
+        template<typename TElem, typename TDim, typename TIdx, typename TDev>
+        class BufGenericSycl;
+
+        template<typename TDev>
+        class QueueGenericSyclBlocking;
+
+        template<typename TDev>
+        class QueueGenericSyclNonBlocking;
+
+        //! The SYCL device handle.
+        template<typename TPltf>
+        class DevGenericSycl
+            : public concepts::Implements<ConceptCurrentThreadWaitFor, DevGenericSycl<TPltf>>
+            , public concepts::Implements<ConceptDev, DevGenericSycl<TPltf>>
+        {
+        public:
+            DevGenericSycl(sycl::device device, sycl::context context) : m_device{device}, m_context{context}
+            {
+            }
+
+            auto operator==(DevGenericSycl const& rhs) const
+            {
+                return (rhs.m_device == m_device);
+            }
+
+            auto operator!=(DevGenericSycl const& rhs) const
+            {
+                return !operator==(rhs);
+            }
+
+            sycl::device m_device;
+            sycl::context m_context;
+            std::vector<sycl::event> m_dependencies = {};
+            std::shared_ptr<std::shared_mutex> mutable mutex_ptr{std::make_shared<std::shared_mutex>()};
+        };
+    } // namespace experimental
+
+    namespace traits
+    {
+        //! The SYCL device name get trait specialization.
+        template<typename TPltf>
+        struct GetName<experimental::DevGenericSycl<TPltf>>
+        {
+            static auto getName(experimental::DevGenericSycl<TPltf> const& dev) -> std::string
+            {
+                return dev.m_device.template get_info<sycl::info::device::name>();
+            }
+        };
+
+        //! The SYCL device available memory get trait specialization.
+        template<typename TPltf>
+        struct GetMemBytes<experimental::DevGenericSycl<TPltf>>
+        {
+            static auto getMemBytes(experimental::DevGenericSycl<TPltf> const& dev) -> std::size_t
+            {
+                return dev.m_device.template get_info<sycl::info::device::global_mem_size>();
+            }
+        };
+
+        //! The SYCL device free memory get trait specialization.
+        template<typename TPltf>
+        struct GetFreeMemBytes<experimental::DevGenericSycl<TPltf>>
+        {
+            static_assert(!sizeof(TPltf), "Querying free device memory not supported for SYCL devices.");
+
+            static auto getFreeMemBytes(experimental::DevGenericSycl<TPltf> const& dev) -> std::size_t
+            {
+            }
+        };
+
+        //! The SYCL device warp size get trait specialization.
+        template<typename TPltf>
+        struct GetWarpSize<experimental::DevGenericSycl<TPltf>>
+        {
+            static auto getWarpSize(experimental::DevGenericSycl<TPltf> const& dev) -> std::size_t
+            {
+                const auto sizes = dev.m_device.template get_info<sycl::info::device::sub_group_sizes>();
```

Comment:
Okay, this poses an interesting problem. With SYCL this can report multiple values. On my Intel GPU for example the warp size can be 8, 16 and 32 and is configurable by the user. The compiler will turn this into one of the corresponding SIMD instructions supported by the GPU.

Maybe this trait should in general return a vector of supported warp sizes @psychocoderHPC and @bernhardmgruber?

Reply:
Hard to say what is best here. Generally for SIMD, the bigger the better. If your kernel threads have high branch divergence though, using a lower warp size might be better.